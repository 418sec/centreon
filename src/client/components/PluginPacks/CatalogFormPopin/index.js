export { default as CatalogFormPopin } from './CatalogFormPopinContainer';