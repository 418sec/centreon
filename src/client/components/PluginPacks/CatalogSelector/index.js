export { default as CatalogSelector } from './CatalogSelectorContainer';