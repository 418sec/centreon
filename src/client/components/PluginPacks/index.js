export { default as PluginPacksList } from './PluginPacksListContainer';