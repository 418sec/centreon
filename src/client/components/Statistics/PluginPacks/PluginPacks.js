import React from 'react'
import { withStyles } from 'material-ui/styles'
import { Bar } from 'react-chartjs-2'
import Typography from 'material-ui/Typography'
import TextField from 'material-ui/TextField'
import { CircularProgress } from 'material-ui/Progress'

const styles = theme => ({
  statisticsContainer: {
    display: 'flex',
    flexDirection: 'column'
  },
  statisticsHeader: {
    display: 'flex',
    flex: '0',
    flexDirection: 'row'
  },
  chartContainer: {
    position: 'relative',
    flex: 1,
    height: '85vh',
    width: '95%'
  },
  title: {
    flex: 1,
  },
  searchField: {
    paddingRight: '100px'
  },
  progress: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    zIndex: 1,
  },

})

const PluginPacks = ({ classes, slug, loading, options, data, searchValue, handleClick, handleSearch }) => (
  <div className={classes.statisticsContainer}>
    <div className={classes.statisticsHeader}>
      <Typography type='display1' align='center' className={classes.title}>
        {data.title}
      </Typography>
      {loading &&
        <CircularProgress className={classes.progress} size={50}/>
      }
      {!slug &&
        <TextField
          id="searchPP"
          label="Search"
          className={classes.searchField}
          value={searchValue}
          onChange={handleSearch}
          margin="normal"
        />
      }
    </div>
    <div className={classes.chartContainer}>
      <Bar
        //height={600}
        data={data}
        getElementAtEvent={handleClick}
        options={options}
      />
    </div>
  </div>
)

export default withStyles(styles)(PluginPacks)