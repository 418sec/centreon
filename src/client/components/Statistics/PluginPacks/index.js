export { default as PluginPacks } from './PluginPacksContainer';