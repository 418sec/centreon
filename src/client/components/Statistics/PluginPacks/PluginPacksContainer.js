import React, { Component } from 'react'
import { connect } from 'react-redux'
import { HeaderBar } from 'Components/HeaderBar'
import PluginPacks from './PluginPacks'
import { getPluginPacksStatistics, getPluginPackStatistics } from 'Services/hubApi'

export class PluginPacksContainer extends Component {

  constructor (props) {
    super(props)

    this.state = {
      searchValue: '',
      options: {
        tooltips: {
          mode: 'index'
        },
        maintainAspectRatio: false,
        responsive: true,
        scales: {
          xAxes: [{
            stacked: true,
            ticks: {
              minRotation: 0,
              maxRotation: 45,
              autoSkip: true,
              maxTicksLimit: 11
            }
          }],
          yAxes: [{
            stacked: true,
            ticks: {
              beginAtZero:true
            }
          }]
        }
      },
      data: {
        labels: [],
        datasets: [
          {
            label: 'Installation',
            backgroundColor: 'rgba(142,204,0,0.2)',
            borderColor: 'rgba(39,101,26,1)',
            borderWidth: 1,
            hoverBackgroundColor: 'rgba(142,204,0,0.4)',
            hoverBorderColor: 'rgba(39,101,26,1)',
            data: [],
          },
          {
            label: 'Update',
            backgroundColor: 'rgba(82,172,230,0.2)',
            borderColor: 'rgba(16,6,159,1)',
            borderWidth: 1,
            hoverBackgroundColor: 'rgba(82,172,230,0.4)',
            hoverBorderColor: 'rgba(16,6,159,1)',
            data: [],
          },
        ]
      }
    }
  }

  componentDidMount = () => {
    this.getPluginPacksStatistics()
  }

  componentWillMount = () => {
    this.timer = null
  }

  getPluginPacksStatistics = () => {
    const { token, getPluginPacksStatistics } = this.props
    const { searchValue } = this.state
    getPluginPacksStatistics(token, searchValue)
  }

  getPluginPackStatistics = (slug) => {
    const { token, getPluginPackStatistics} = this.props
    getPluginPackStatistics(token, slug)
  }

  formatPluginPacksStatistics = () => {
    const { pluginPacksStatistics } = this.props
    const { data } = this.state

    let labels = []
    let installation = []
    let update = []

    for (const slug of Object.keys(pluginPacksStatistics)) {
      labels.push(slug)
      installation.push(pluginPacksStatistics[slug].totalInstalled)
      update.push(pluginPacksStatistics[slug].totalUpdated)
    }

    data.title = 'All plugin packs'
    data.labels = labels
    data.datasets[0].data = installation
    data.datasets[1].data = update

    return data
  }

  formatPluginPackStatistics = () => {
    const { slug, pluginPacksStatistics } = this.props
    const { data } = this.state

    let labels = []
    let installation = []
    let update = []

    for (const version of Object.keys(pluginPacksStatistics[slug].versions)) {
      labels.push(version)
      installation.push(pluginPacksStatistics[slug].versions[version].totalInstalled)
      update.push(pluginPacksStatistics[slug].versions[version].totalUpdated)
    }

    data.title = slug
    data.labels = labels
    data.datasets[0].data = installation
    data.datasets[1].data = update

    return data
  }

  handleSearch = (event) => {
    const value = event.target.value

    this.setState({ searchValue: value })

    clearTimeout(this.timer)

    this.timer = setTimeout(
      () => {
        this.getPluginPacksStatistics()
      },
      500
    )
  }

  handleClick = (element) => {
    const { loading, slug } = this.props

    // Avoid to fetch multiple times on click spam
    if (!loading) {
      if (!element[0] || slug) { // Fetch all plugin packs statistics if click outside data or if a single plugin pack is already displayed
        this.getPluginPacksStatistics()
      } else { // Fetch single plugin pack statistics
        this.getPluginPackStatistics(element[0]._model.label)
      }
    }
  }

  render = () => {
    const { loading, slug } = this.props
    const { options, searchValue } = this.state

    let formattedData
    if (!slug) {
      formattedData = this.formatPluginPacksStatistics()
    } else {
      formattedData = this.formatPluginPackStatistics()
    }

    return (
      <PluginPacks
        slug={slug}
        searchValue={searchValue}
        handleSearch={this.handleSearch}
        handleClick={this.handleClick}
        loading={loading}
        options={options}
        data={formattedData}
      />
    )
  }
}

const mapStateToProps = (store) => {
  return {
    token: store.authReducer.token,
    loading: store.statisticsReducer.loading,
    slug: store.statisticsReducer.slug,
    pluginPacksStatistics: store.statisticsReducer.pluginPacks,
  }
}


const mapDispatchToProps = (dispatch) => {
  return {
    getPluginPacksStatistics: (token, search) => {
      return dispatch(getPluginPacksStatistics(token, search))
    },
    getPluginPackStatistics: (token, slug) => {
      return dispatch(getPluginPackStatistics(token, slug))
    },
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(PluginPacksContainer)