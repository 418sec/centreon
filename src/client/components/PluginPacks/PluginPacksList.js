import React, { Component } from 'react'
import {
  PagingState,
  FilteringState,
  CustomPaging
} from '@devexpress/dx-react-grid'
import {
  Grid,
  Table,
  TableHeaderRow,
  TableFilterRow,
  PagingPanel
} from '@devexpress/dx-react-grid-material-ui'
import PropTypes from 'prop-types'
import { withStyles } from 'material-ui/styles'
import { TableRow, TableCell } from 'material-ui'
import { HeaderBar } from 'Components/HeaderBar'
import { CatalogFormPopin } from 'Components/PluginPacks/CatalogFormPopin'
import { CatalogSelector } from 'Components/PluginPacks/CatalogSelector'
import { ReleasedButton } from 'Components/PluginPacks/ReleasedButton'
import { FilterCell } from 'Components/PluginPacks/FilterCell'

const styles = theme => ({
  root: {
    padding: '0 15px',
    overflow: 'auto'
  },
  TableCell: {
    wordWrap: 'break-word',
    padding: '4px 14px',
  },
  check: {
    fill: '#88B917'
  },
  checkInactive: {
    fill: '#d1d2d4'
  },
  textField: {
    '&:hover': {
      background:'transparent'
    },
  }
})

class PluginPacksList extends Component {

  tableRowTemplate = ({children, row}) => {
    return (
      <TableRow id={row.slug}>
        {children}
      </TableRow>
    )
  }

  render = () => {
    const { columns, rows, itemsPerPage, totalItems, currentPage, classes } = this.props

    return (
      <section>
        <div id="pluginPackList" className={classes.root}>
          <Grid
            columns={columns}
            rows={rows}
          >
            <FilteringState
              onFiltersChange={this.props.changeFilters}
            />
            <PagingState
              currentPage={currentPage}
              onCurrentPageChange={this.props.changeCurrentPage}
              pageSize={itemsPerPage}
            />
            <CustomPaging
              totalCount={totalItems}
            />
            <Table
              rowComponent={this.tableRowTemplate}
              cellComponent={({ column, row, value }) => {
                if (column.name === 'released') {
                  return (
                    <TableCell>
                      <ReleasedButton
                        pluginPackId={row.id}
                      />
                    </TableCell>
                  )
                } else if (column.name === 'catalog_level') {
                  return (
                    <TableCell>
                      <CatalogSelector
                        id={row.id}
                        level={value}
                      />
                    </TableCell>
                  )
                }

                return (
                  <TableCell className={classes.TableCell}>{value}</TableCell>
                )
              }}
            />
            <TableHeaderRow />
            <PagingPanel id={"pagingPanel"}/>
            <TableFilterRow
              messages={{ filterPlaceholder: 'Filter' }}
              cellComponent={FilterCell}
            />
          </Grid>
        </div>
        <CatalogFormPopin />
      </section>
    )
  }
}

/**
 * Instance list Prototypes
 */
PluginPacksList.propTypes = {
  /**
   * Instance list css classes
   * @param {Object} classes
   */
  classes: PropTypes.object.isRequired,
}

PluginPacksList.defaultProps = {
  currentPage: 1
}

export default withStyles(styles)(PluginPacksList)
