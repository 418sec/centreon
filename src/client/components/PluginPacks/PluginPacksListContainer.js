import React, { Component } from 'react'
import { withRouter } from 'react-router-dom'
import { connect } from 'react-redux'
import PluginPacksList from './PluginPacksList'
import { getPluginPacks } from 'Services/hubApi'
import { cleanHeader, replaceTitle} from 'Redux/header/headerActions'

class PluginPacksListContainer extends Component {

  constructor (props) {
    super(props)

    this.state = {
      columns: [
        {name: 'slug', title: 'Slug'},
        {name: 'name', title: 'Name'},
        {name: 'version', title: 'Version'},
        {name: 'catalog_level', title: 'Catalog'},
        {name: 'released', title: 'Released'},
      ],
      currentPage: 0,
      itemsPerPage: 10,
      filters: []
    }
  }

  componentDidMount = () => {
    const { cleanHeader, replaceHeaderTitle } = this.props
    cleanHeader()
    replaceHeaderTitle('Plugin Packs')

    this.getData()
  }

  changeFilters = (filters) => {
    this.setState(
      {
        filters: filters,
        currentPage: 0
      },
      this.getData
    )
  }

  changeCurrentPage = (currentPage) => {
    this.setState(
      {currentPage: currentPage},
      this.getData
    )
  }

  /**
   * example : page[number]=1&page[size]=30
   * @returns {string|*}
   */
  getPagination = () => {
    const { currentPage, itemsPerPage } = this.state
    return `?page[number]=${currentPage + 1}&page[size]=${itemsPerPage}`
  }

  /**
   * example : filter[instanceName]=centreon&filter[fingerprint]=abc
   * @returns {string|*}
   */
  getFilters = () => {
    const { filters } = this.state;

    return filters.reduce((acc, { columnName, value }) => {
      if (columnName == 'catalog_level' && value == '0') {
        acc.push(`filter[with_catalog]=false`)
      } else {
        acc.push(`filter[${columnName}]=${value}`)
      }
      return acc
    }, []).join('&')
  }

  getData = () => {
    const { token, displayPluginPacks } = this.props
    displayPluginPacks(this.getPagination(), this.getFilters(), token)
  }

  render = () => {
    const { columns, currentPage } = this.state
    const { items, itemsPerPage, totalPages, totalItems, ...props } = this.props

    const rows = items ?
      Object.keys(items).reduce((acc, key) => {
        acc[items[key].order] = items[key]
        return acc
      }, [])
      : []

    return (
      <PluginPacksList
        columns={columns}
        rows={rows}
        itemsPerPage={itemsPerPage}
        totalPages={totalPages}
        totalItems={totalItems}
        currentPage={currentPage}
        changeFilters={this.changeFilters}
        changeCurrentPage={this.changeCurrentPage}
        {...props}
      />
    )
  }
}

const mapStateToProps = (store) => {
  return {
    token: store.authReducer.token,
    items: store.pluginPacksReducer.items,
    totalPages: store.pluginPacksReducer.totalPages,
    totalItems: store.pluginPacksReducer.totalItems,
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    cleanHeader: () => {
      return dispatch(cleanHeader())
    },
    replaceHeaderTitle: (title) => {
      return dispatch(replaceTitle(title))
    },
    displayPluginPacks: (pagination, filters, token) => {
      return dispatch(getPluginPacks(pagination, filters, token))
    },
  }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(PluginPacksListContainer))