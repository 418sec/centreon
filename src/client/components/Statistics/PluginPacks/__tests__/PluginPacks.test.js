/* eslint-env jest */
import React from 'react'
import Enzyme from 'enzyme'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16'
import { PluginPacksContainer as PluginPacks } from 'Components/Statistics/PluginPacks/PluginPacksContainer'
import configureStore from 'redux-mock-store'

Enzyme.configure({ adapter: new Adapter() })

describe('Plugin Packs Statistics', () => {
  const mockStore = configureStore()
  const getPluginPacksStatistics = jest.fn()
  const getPluginPackStatistics = jest.fn()
  const props = {
    getPluginPacksStatistics: jest.fn(),
    getPluginPackStatistics: jest.fn(),
    token: '',
    loading: false,
    slug: 'pp2',
    pluginPacksStatistics: {
      'pp1': {
        totalInstalled: 17,
        totalUpdated: 5,
        versions: {
          '1.0.0': {
              totalInstalled: 3,
              totalUpdated: 0
          }
        }
      },
      'pp2': {
        totalInstalled: 7,
        totalUpdated: 1,
        versions: {
          '1.0.0': {
            totalInstalled: 1,
            totalUpdated: 0
          },
          '1.0.1': {
            totalInstalled: 6,
            totalUpdated: 1
          }
        }
      }
    }
  }
  let component

  it('renders correctly', () => {
    component = shallow(
      <PluginPacks
        {...props}
      />
    ).dive().dive()
    expect(toJson(component)).toMatchSnapshot()
  })

  it('formats plugin packs stats correctly', () => {
    let stats
    const expectedTitle = 'All plugin packs',
      expectedLabels = [ 'pp1', 'pp2' ],
      expectedData1 = [
        props.pluginPacksStatistics['pp1'].totalInstalled,
        props.pluginPacksStatistics['pp2'].totalInstalled,
      ],
      expectedData2 = [
        props.pluginPacksStatistics['pp1'].totalUpdated,
        props.pluginPacksStatistics['pp2'].totalUpdated,
      ]

    component = mount(
      <PluginPacks
        {...props}
      />
    )

    stats = component.instance().formatPluginPacksStatistics()

    expect(stats.title).toEqual(expectedTitle)
    expect(stats.labels).toEqual(expectedLabels)
    expect(stats.datasets[0].data).toEqual(expectedData1)
    expect(stats.datasets[1].data).toEqual(expectedData2)
  })

  it('formats plugin pack stats correctly', () => {
    let stats
    const expectedTitle = 'pp2',
      expectedLabels = [ '1.0.0', '1.0.1' ],
      expectedData1 = [
        props.pluginPacksStatistics['pp2'].versions['1.0.0'].totalInstalled,
        props.pluginPacksStatistics['pp2'].versions['1.0.1'].totalInstalled,
      ],
      expectedData2 = [
        props.pluginPacksStatistics['pp2'].versions['1.0.0'].totalUpdated,
        props.pluginPacksStatistics['pp2'].versions['1.0.1'].totalUpdated,
      ]


    component = mount(
      <PluginPacks
        {...props}
      />
    )

    stats = component.instance().formatPluginPackStatistics()

    expect(stats.title).toEqual(expectedTitle)
    expect(stats.labels).toEqual(expectedLabels)
    expect(stats.datasets[0].data).toEqual(expectedData1)
    expect(stats.datasets[1].data).toEqual(expectedData2)
  })
})