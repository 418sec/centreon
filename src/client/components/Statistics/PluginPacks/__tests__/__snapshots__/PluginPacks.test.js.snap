// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugin Packs Statistics renders correctly 1`] = `
<div
  className="PluginPacks-statisticsContainer-1"
>
  <div
    className="PluginPacks-statisticsHeader-2"
  >
    <WithStyles(Typography)
      align="center"
      className="PluginPacks-title-4"
      type="display1"
    >
      pp2
    </WithStyles(Typography)>
  </div>
  <div
    className="PluginPacks-chartContainer-3"
  >
    <Bar
      data={
        Object {
          "datasets": Array [
            Object {
              "backgroundColor": "rgba(142,204,0,0.2)",
              "borderColor": "rgba(39,101,26,1)",
              "borderWidth": 1,
              "data": Array [
                1,
                6,
              ],
              "hoverBackgroundColor": "rgba(142,204,0,0.4)",
              "hoverBorderColor": "rgba(39,101,26,1)",
              "label": "Installation",
            },
            Object {
              "backgroundColor": "rgba(82,172,230,0.2)",
              "borderColor": "rgba(16,6,159,1)",
              "borderWidth": 1,
              "data": Array [
                0,
                1,
              ],
              "hoverBackgroundColor": "rgba(82,172,230,0.4)",
              "hoverBorderColor": "rgba(16,6,159,1)",
              "label": "Update",
            },
          ],
          "labels": Array [
            "1.0.0",
            "1.0.1",
          ],
          "title": "pp2",
        }
      }
      getElementAtEvent={[Function]}
      options={
        Object {
          "maintainAspectRatio": false,
          "responsive": true,
          "scales": Object {
            "xAxes": Array [
              Object {
                "stacked": true,
                "ticks": Object {
                  "autoSkip": true,
                  "maxRotation": 45,
                  "maxTicksLimit": 11,
                  "minRotation": 0,
                },
              },
            ],
            "yAxes": Array [
              Object {
                "stacked": true,
                "ticks": Object {
                  "beginAtZero": true,
                },
              },
            ],
          },
          "tooltips": Object {
            "mode": "index",
          },
        }
      }
    />
  </div>
</div>
`;
